File: packages/sdk/src/harness/loop/worktree_manager.py
"""
Worktree manager for git worktree lifecycle management.

This module provides WorktreeManager for creating, tracking, and cleaning up
git worktrees for parallel goal execution.
"""

from __future__ import annotations

import asyncio
import logging
import os
import subprocess
from pathlib import Path
from typing import TYPE_CHECKING

from harness.loop.worktree_types import WORKTREES_DIR, WorktreeError

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)


class WorktreeManager:
    """
    Git worktree lifecycle manager.

    Manages creation, tracking, and cleanup of git worktrees for
    parallel goal execution in isolated environments.

    Features:
    - Async git command execution
    - Orphan worktree recovery on startup
    - Thread-safe operation tracking

    Example:
        ```python
        manager = WorktreeManager("/path/to/repo")

        # Create worktree
        path, branch = await manager.create_worktree(
            name="feature-auth",
            base_branch="main",
            create_branch=True,
        )

        # List active worktrees
        for name in manager.list_worktrees():
            print(f"Active: {name}")

        # Cleanup
        await manager.cleanup_worktree("feature-auth")
        ```
    """

    def __init__(self, repo_root: str):
        """
        Initialize WorktreeManager.

        Args:
            repo_root: Path to the main git repository root

        Raises:
            WorktreeError: If repo_root is not a valid git repository
        """
        self.repo_root = os.path.abspath(repo_root)
        self._worktrees: dict[str, str] = {}  # name -> path

        # Verify this is a git repository
        if not os.path.isdir(os.path.join(self.repo_root, ".git")):
            # Check if it's a worktree itself (git dir is in .git file)
            git_file = os.path.join(self.repo_root, ".git")
            if not os.path.isfile(git_file):
                raise WorktreeError(f"Not a git repository: {self.repo_root}")

        # Recover orphaned worktrees from previous runs
        self._recover_orphaned_worktrees()

    def _recover_orphaned_worktrees(self) -> None:
        """
        Recover orphaned worktrees from previous runs.

        Scans git worktree list and rebuilds internal tracking for
        worktrees created by this system (identified by path prefix).

        This provides resilience against process crashes - the system
        can recover state and clean up properly on restart.
        """
        try:
            result = subprocess.run(
                ["git", "worktree", "list", "--porcelain"],
                cwd=self.repo_root,
                capture_output=True,
                text=True,
                check=True,
            )

            # Exact prefix matching to avoid recovering unrelated worktrees
            expected_prefix = f"{self.repo_root}/{WORKTREES_DIR}/"

            current_path = None
            for line in result.stdout.splitlines():
                if line.startswith("worktree "):
                    current_path = line.split(" ", 1)[1]
                elif line.startswith("branch ") and current_path:
                    # Only recover if path matches our expected prefix
                    if current_path.startswith(expected_prefix):
                        name = Path(current_path).name
                        self._worktrees[name] = current_path
                        logger.info(f"Recovered orphan worktree: {name}")
                    current_path = None

            if self._worktrees:
                logger.info(
                    f"Recovered {len(self._worktrees)} orphan worktree(s)"
                )

        except subprocess.CalledProcessError as e:
            logger.warning(f"Failed to list worktrees for recovery: {e}")
        except Exception as e:
            logger.warning(f"Error during worktree recovery: {e}")

    async def create_worktree(
        self,
        name: str,
        base_branch: str,
        create_branch: bool = True,
    ) -> tuple[str, str]:
        """
        Create a git worktree for isolated execution.

        Args:
            name: Unique name for this worktree
            base_branch: Base branch to create from
            create_branch: If True, create a new branch named `name`;
                          if False, checkout base_branch in worktree

        Returns:
            Tuple of (worktree_path, branch_name)

        Raises:
            WorktreeError: If worktree creation fails
        """
        # Check if worktree already exists
        if name in self._worktrees:
            raise WorktreeError(f"Worktree already exists: {name}")

        # Determine branch name
        branch_name = name if create_branch else base_branch

        # Build worktree path
        worktree_path = f"{self.repo_root}/{WORKTREES_DIR}/{name}"

        # Build git worktree add command
        cmd = ["git", "worktree", "add"]

        if create_branch:
            cmd.extend(["-b", branch_name])

        cmd.extend([worktree_path, base_branch])

        logger.debug(f"Creating worktree: {' '.join(cmd)}")

        # Execute git command
        proc = await asyncio.create_subprocess_exec(
            *cmd,
            cwd=self.repo_root,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await proc.communicate()

        if proc.returncode != 0:
            error_msg = stderr.decode().strip() or stdout.decode().strip()
            raise WorktreeError(f"Failed to create worktree '{name}': {error_msg}")

        # Track the worktree
        self._worktrees[name] = worktree_path
        logger.info(f"Created worktree: {name} at {worktree_path}")

        return worktree_path, branch_name

    async def cleanup_worktree(self, name: str, force: bool = False) -> bool:
        """
        Remove a git worktree.

        Args:
            name: Name of the worktree to remove
            force: If True, force removal even with uncommitted changes

        Returns:
            True if cleanup succeeded, False if worktree not found

        Raises:
            WorktreeError: If cleanup fails
        """
        path = self._worktrees.get(name)
        if not path:
            logger.warning(f"Worktree not found for cleanup: {name}")
            return False

        # Build git worktree remove command
        cmd = ["git", "worktree", "remove", path]
        if force:
            cmd.append("--force")

        logger.debug(f"Removing worktree: {' '.join(cmd)}")

        proc = await asyncio.create_subprocess_exec(
            *cmd,
            cwd=self.repo_root,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await proc.communicate()

        if proc.returncode != 0:
            error_msg = stderr.decode().strip()
            if force:
                # Try more aggressive cleanup
                logger.warning(
                    f"Force cleanup for worktree {name}: {error_msg}"
                )
                # Use git worktree prune
                await self._prune_worktrees()

        # Remove from tracking
        if name in self._worktrees:
            del self._worktrees[name]

        logger.info(f"Cleaned up worktree: {name}")
        return True

    async def _prune_worktrees(self) -> None:
        """
        Prune stale worktree references.

        Called when force cleanup is needed to clean up git's internal
        worktree tracking.
        """
        proc = await asyncio.create_subprocess_exec(
            "git",
            "worktree",
            "prune",
            cwd=self.repo_root,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        await proc.communicate()

    def list_worktrees(self) -> list[str]:
        """
        List all tracked worktrees.

        Returns:
            List of worktree names
        """
        return list(self._worktrees.keys())

    def get_worktree_path(self, name: str) -> str | None:
        """
        Get the path to a worktree.

        Args:
            name: Worktree name

        Returns:
            Path to the worktree, or None if not found
        """
        return self._worktrees.get(name)

    async def get_commit_count(
        self,
        name: str,
        base_branch: str = "main",
    ) -> int:
        """
        Get the number of commits in a worktree branch vs base branch.

        Args:
            name: Worktree name
            base_branch: Base branch to compare against

        Returns:
            Number of commits ahead of base branch
        """
        path = self._worktrees.get(name)
        if not path:
            return 0

        try:
            proc = await asyncio.create_subprocess_exec(
                "git",
                "rev-list",
                "--count",
                f"{base_branch}..HEAD",
                cwd=path,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            stdout, _ = await proc.communicate()

            if proc.returncode == 0:
                return int(stdout.decode().strip())

        except (ValueError, subprocess.CalledProcessError):
            pass

        return 0

    async def is_dirty(self) -> bool:
        """
        Check if the main repository has uncommitted changes.

        Returns:
            True if there are uncommitted changes
        """
        proc = await asyncio.create_subprocess_exec(
            "git",
            "status",
            "--porcelain",
            cwd=self.repo_root,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, _ = await proc.communicate()

        return bool(stdout.decode().strip())

    async def cleanup_all(self, force: bool = False) -> int:
        """
        Clean up all tracked worktrees.

        Args:
            force: If True, force cleanup

        Returns:
            Number of worktrees cleaned up
        """
        count = 0
        for name in list(self._worktrees.keys()):
            try:
                if await self.cleanup_worktree(name, force=force):
                    count += 1
            except WorktreeError as e:
                logger.warning(f"Failed to cleanup worktree {name}: {e}")

        return count

    def __repr__(self) -> str:
        """String representation."""
        return f"WorktreeManager(repo_root='{self.repo_root}', worktrees={len(self._worktrees)})"

File: packages/sdk-java/harness-sdk-loop/src/main/java/com/harness/loop/worktree/WorktreeManager.java
package com.harness.loop.worktree;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CompletableFuture;

/**
 * Git worktree lifecycle manager.
 *
 * <p>Manages creation, tracking, and cleanup of git worktrees for
 * parallel goal execution in isolated environments.</p>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * WorktreeManager manager = new WorktreeManager("/path/to/repo");
 *
 * // Create worktree
 * WorktreeInfo info = manager.createWorktree("feature-auth", "main", true).join();
 *
 * // List active worktrees
 * for (String name : manager.listWorktrees()) {
 *     System.out.println("Active: " + name);
 * }
 *
 * // Cleanup
 * manager.cleanupWorktree("feature-auth").join();
 * }</pre>
 */
public class WorktreeManager {
    private static final Logger logger = LoggerFactory.getLogger(WorktreeManager.class);

    private final String repoRoot;
    private final Map<String, String> worktrees = new ConcurrentHashMap<>();

    /**
     * Create a new WorktreeManager.
     *
     * @param repoRoot Path to the main git repository root
     * @throws WorktreeError If repoRoot is not a valid git repository
     */
    public WorktreeManager(String repoRoot) {
        this.repoRoot = Paths.get(repoRoot).toAbsolutePath().toString();

        // Verify this is a git repository
        Path gitDir = Paths.get(this.repoRoot, ".git");
        if (!Files.isDirectory(gitDir) && !Files.isRegularFile(gitDir)) {
            throw new WorktreeError("Not a git repository: " + this.repoRoot);
        }

        // Recover orphaned worktrees from previous runs
        recoverOrphanedWorktrees();
    }

    /**
     * Recover orphaned worktrees from previous runs.
     */
    private void recoverOrphanedWorktrees() {
        try {
            ProcessBuilder pb = new ProcessBuilder("git", "worktree", "list", "--porcelain");
            pb.directory(new File(repoRoot));
            pb.redirectErrorStream(true);
            Process process = pb.start();

            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            String currentPath = null;
            String expectedPrefix = repoRoot + "/" + WorktreeConfig.WORKTREES_DIR + "/";

            while ((line = reader.readLine()) != null) {
                if (line.startsWith("worktree ")) {
                    currentPath = line.substring("worktree ".length());
                } else if (line.startsWith("branch ") && currentPath != null) {
                    if (currentPath.startsWith(expectedPrefix)) {
                        String name = Paths.get(currentPath).getFileName().toString();
                        worktrees.put(name, currentPath);
                        logger.info("Recovered orphan worktree: {}", name);
                    }
                    currentPath = null;
                }
            }

            process.waitFor();

            if (!worktrees.isEmpty()) {
                logger.info("Recovered {} orphan worktree(s)", worktrees.size());
            }

        } catch (Exception e) {
            logger.warn("Failed to list worktrees for recovery: {}", e.getMessage());
        }
    }

    /**
     * Create a git worktree for isolated execution.
     *
     * @param name Unique name for this worktree
     * @param baseBranch Base branch to create from
     * @param createBranch If true, create a new branch named {@code name}
     * @return CompletableFuture with WorktreeInfo
     */
    public CompletableFuture<WorktreeInfo> createWorktree(String name, String baseBranch, boolean createBranch) {
        return CompletableFuture.supplyAsync(() -> {
            // Check if worktree already exists
            if (worktrees.containsKey(name)) {
                throw new WorktreeError("Worktree already exists: " + name);
            }

            String branchName = createBranch ? name : baseBranch;
            String worktreePath = repoRoot + "/" + WorktreeConfig.WORKTREES_DIR + "/" + name;

            // Build git worktree add command
            List<String> cmd = new ArrayList<>();
            cmd.add("git");
            cmd.add("worktree");
            cmd.add("add");

            if (createBranch) {
                cmd.add("-b");
                cmd.add(branchName);
            }

            cmd.add(worktreePath);
            cmd.add(baseBranch);

            logger.debug("Creating worktree: {}", String.join(" ", cmd));

            try {
                ProcessBuilder pb = new ProcessBuilder(cmd);
                pb.directory(new File(repoRoot));
                pb.redirectErrorStream(true);
                Process process = pb.start();

                int exitCode = process.waitFor();

                if (exitCode != 0) {
                    String output = readProcessOutput(process);
                    throw new WorktreeError("Failed to create worktree '" + name + "': " + output);
                }

                worktrees.put(name, worktreePath);
                logger.info("Created worktree: {} at {}", name, worktreePath);

                return new WorktreeInfo(worktreePath, branchName);

            } catch (IOException | InterruptedException e) {
                throw new WorktreeError("Failed to create worktree '" + name + "'", e);
            }
        });
    }

    /**
     * Remove a git worktree.
     *
     * @param name Name of the worktree to remove
     * @param force If true, force removal even with uncommitted changes
     * @return CompletableFuture with true if cleanup succeeded
     */
    public CompletableFuture<Boolean> cleanupWorktree(String name, boolean force) {
        return CompletableFuture.supplyAsync(() -> {
            String path = worktrees.get(name);
            if (path == null) {
                logger.warn("Worktree not found for cleanup: {}", name);
                return false;
            }

            List<String> cmd = new ArrayList<>();
            cmd.add("git");
            cmd.add("worktree");
            cmd.add("remove");
            cmd.add(path);

            if (force) {
                cmd.add("--force");
            }

            logger.debug("Removing worktree: {}", String.join(" ", cmd));

            try {
                ProcessBuilder pb = new ProcessBuilder(cmd);
                pb.directory(new File(repoRoot));
                pb.redirectErrorStream(true);
                Process process = pb.start();

                int exitCode = process.waitFor();

                if (exitCode != 0 && force) {
                    // Try prune
                    logger.warn("Force cleanup failed for {}, trying prune", name);
                    pruneWorktrees();
                }

                worktrees.remove(name);
                logger.info("Cleaned up worktree: {}", name);
                return true;

            } catch (IOException | InterruptedException e) {
                logger.warn("Failed to cleanup worktree {}: {}", name, e.getMessage());
                return false;
            }
        });
    }

    /**
     * Prune stale worktree references.
     */
    private void pruneWorktrees() {
        try {
            ProcessBuilder pb = new ProcessBuilder("git", "worktree", "prune");
            pb.directory(new File(repoRoot));
            Process process = pb.start();
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            logger.warn("Failed to prune worktrees: {}", e.getMessage());
        }
    }

    /**
     * List all tracked worktrees.
     */
    public List<String> listWorktrees() {
        return new ArrayList<>(worktrees.keySet());
    }

    /**
     * Get the path to a worktree.
     */
    public String getWorktreePath(String name) {
        return worktrees.get(name);
    }

    /**
     * Get the number of commits in a worktree branch vs base branch.
     */
    public CompletableFuture<Integer> getCommitCount(String name, String baseBranch) {
        return CompletableFuture.supplyAsync(() -> {
            String path = worktrees.get(name);
            if (path == null) {
                return 0;
            }

            try {
                ProcessBuilder pb = new ProcessBuilder(
                        "git", "rev-list", "--count", baseBranch + "..HEAD");
                pb.directory(new File(path));
                pb.redirectErrorStream(true);
                Process process = pb.start();

                BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
                String line = reader.readLine().trim();
                process.waitFor();

                return Integer.parseInt(line);

            } catch (IOException | InterruptedException | NumberFormatException e) {
                return 0;
            }
        });
    }

    /**
     * Check if the main repository has uncommitted changes.
     */
    public CompletableFuture<Boolean> isDirty() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                ProcessBuilder pb = new ProcessBuilder("git", "status", "--porcelain");
                pb.directory(new File(repoRoot));
                pb.redirectErrorStream(true);
                Process process = pb.start();

                BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
                String line = reader.readLine();
                process.waitFor();

                return line != null && !line.isEmpty();

            } catch (IOException | InterruptedException e) {
                return false;
            }
        });
    }

    /**
     * Clean up all tracked worktrees.
     */
    public CompletableFuture<Integer> cleanupAll(boolean force) {
        List<CompletableFuture<Boolean>> futures = new ArrayList<>();

        for (String name : new ArrayList<>(worktrees.keySet())) {
            futures.add(cleanupWorktree(name, force));
        }

        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
                .thenApply(v -> (int) futures.stream().filter(f -> f.join()).count());
    }

    private String readProcessOutput(Process process) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line).append("\n");
        }
        return sb.toString().trim();
    }

    /**
     * Info about a created worktree.
     */
    public record WorktreeInfo(String path, String branchName) {}

    @Override
    public String toString() {
        return "WorktreeManager{repoRoot='" + repoRoot + "', worktrees=" + worktrees.size() + "}";
    }
}
